### ISSUE TITLE
Deployment versioning

### ISSUE THREAD

**Issue by j-tr** [PrefectHQ/prefect#10923]
### First check

- [X] I added a descriptive title to this issue.
- [X] I used the GitHub search to find a similar request and didn't find it.
- [X] I searched the Prefect documentation for this feature.

### Prefect Version

2.x

### Describe the current behavior

Deployments of a flow are uniquely identified by name. Therefore, it is currently not possible to create multiple versions of a deployment with the same name. 
Having deployed multiple (historic) versions of a deployment at the same time requires encoding the version in the deployment name. These pseudo-versions are displayed as individual deployments in the UI.

### Describe the proposed behavior

Identify deployments by name and optionally version, defaulting to the latest available version.
Group deployments by name in the UI, with easy access to the latest and all prior versions (via a version dropdown?).

### Example Use

Versioning of Prefect flows becomes important when running flows programmatically via the API from possibly multiple external services that rely on a stable interface of the flow while still having human users that run flows manually. Encoding the deployment version into the deployment name can harm the UX for human users as each version is a completely separate deployment, leading to a large number of deployments in the UI. Grouping multiple versions of a deployment in the UI could mitigate this.

As a service developer, I would like to be able to uniquely identify a deployment by its name and version via the Prefect API and be able to run any historic version of a deployment at any time.

As a human user of Prefect, I mostly need to execute the latest version of a deployment because I can easily adapt to changes in the flow's interface. Therefore, I would like to have the possibility to group deployments per name to keep a clean UI and easily find the latest version of a deployment.

**Comment by desertaxle** [comment]
Thanks for the issue @j-tr! To implement this feature, the user would also need to version the underlying code for the deployment. Do you have a proposed experience for pinning a deployment to a particular code revision?

**Comment by j-tr** [comment]
indeed, I think this would highly depend on where the underlying code is stored.
For git-based storage, this would be no problem. For docker or S3 based storage, do you have a better idea than including the version in the prefix or image uri?
